### PR TITLE
Faces branch on top of master

### PR DESCRIPTION
**Pull request by ricardofantin**
All changes from mine faces branch are included in this single commit.

One of the main problems with the old branch, is the unstable commits that doesn't compile. This single commit changes everything in one step jumping over that problem.

compile with:
meson build/ -Dface=true

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -81,6 +81,11 @@ if get_option('disable-dupe-detection')
   add_global_arguments(['--define=NO_DUPE_DETECTION'], language : vala)
 endif
 
+if get_option('face')
```
Comment (phako) [inline meson.build]: I would suggest something like "faces" or "face-detection"

File: facedetect/shotwell-facedetect.cpp [hunk 2]
```diff
@@ -0,0 +1,125 @@
+#include "opencv2/objdetect/objdetect.hpp"
```
Comment (phako) [inline facedetect/shotwell-facedetect.cpp:10]: Please add the Copyright header and the original author to this.

File: src/db/FaceLocationTable.vala [hunk 3]
```diff
@@ -0,0 +1,202 @@
+/* Copyright 2011-2012 Yorba Foundation
```
Comment (phako) [inline src/db/FaceLocationTable.vala]: Please update according to current copyright

File: src/editing_tools/EditingTools.vala [hunk 4]
```diff
@@ -1815,6 +1815,958 @@ public class CropTool : EditingTool {
...
+        CLICK_TO_EDIT,
+        NOT_EDITING,
+        CREATING_DRAGGING,
+        CREATING_EDITING,
+        EDITING,
+        DETECTING_FACES,
+        DETECTING_FACES_FINISHED
+    }
+
+    public class FaceWidget : Gtk.HBox {
```
Comment (phako) [inline src/editing_tools/EditingTools.vala]: Please fix deprecated Gtk+ widgets such as HBox => Box with orientation: horizontal

File: src/editing_tools/EditingTools.vala [hunk 5]
```diff
@@ -1815,6 +1815,958 @@ public class CropTool : EditingTool {
...
+            attrs_bold = new Pango.AttrList();
+            attrs_bold.insert(Pango.attr_weight_new(Pango.Weight.BOLD));
+            attrs_normal = new Pango.AttrList();
+            attrs_normal.insert(Pango.attr_weight_new(Pango.Weight.NORMAL));
+        }
+
+        public FaceWidget (FaceShape face_shape) {
+            spacing = CONTROL_SPACING;
+
+            edit_button = new Gtk.Button.from_stock(Gtk.Stock.EDIT);
```
Comment (phako) [inline src/editing_tools/EditingTools.vala]: Same for stock buttons, those are deprecated as well

File: src/editing_tools/EditingTools.vala [hunk 6]
```diff
@@ -1815,6 +1815,958 @@ public class CropTool : EditingTool {
...
+                    image_path
+                };
+                Process.spawn_sync(null, argv, null, SpawnFlags.STDERR_TO_DEV_NULL, null, out output);
+                
+            } catch (SpawnError e) {
+                stderr.printf("Error trying to spawn face detection program: %s\n", e.message);
+                assert_not_reached();
+            }
+            
+            faces = new Gee.PriorityQueue<string>();
```
Comment (phako) [inline src/editing_tools/EditingTools.vala]: As a backlog item, we should move this to something that's slightly nicer to parse than this

File: src/editing_tools/EditingTools.vala [hunk 7]
```diff
@@ -1815,6 +1815,958 @@ public class CropTool : EditingTool {
 
 }
 
+#if ENABLE_FACES
```
Comment (phako) [inline src/editing_tools/EditingTools.vala]: I think this is big enough to split it off into its own file

File: src/faces/Face.vala [hunk 8]
```diff
@@ -0,0 +1,682 @@
...
+            bool is_added = faces.add(face);
+            assert(is_added);
+        }
+        
+        base.notify_container_contents_added(container, added, relinking);
+    }
+    
+    protected override void notify_container_contents_removed(ContainerSource container, 
+        Gee.Collection<DataObject> removed, bool unlinking) {
+//        Gee.Collection<DataSource> removed, bool unlinking) {
```
Comment (phako) [inline src/faces/Face.vala]: No commented code - why is that here?

File: src/faces/FacePage.vala [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        Config.Facade.get_instance().get_event_photos_sort(out sort_order, out sort_by);
+    }
+
+    protected override void set_config_photos_sort(bool sort_order, int sort_by) {
+        Config.Facade.get_instance().set_event_photos_sort(sort_order, sort_by);
+    }
+
+    private const GLib.ActionEntry[] entries = {
+        { "DeleteFace", on_delete_face },
+		{ "RenameFace", on_rename_face },
```
Comment (phako) [inline src/faces/FacePage.vala]: Indentation error

File: src/faces/FacePage.vala [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+        map.add_action_entries (entries, this);
+    }
+
+    protected override InjectionGroup[] init_collect_injection_groups() {
+        InjectionGroup[] groups = base.init_collect_injection_groups();
+        
+        //groups.add_menu_item(_("Face Location"), "FaceLocation");
+       
+        //groups += create_photos_faces_injectables();
```
Comment (phako) [inline src/faces/FacePage.vala]: Commented code

File: src/faces/FacePage.vala [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+		groups += create_faces_menu_injectables();
+
```
Comment (phako) [inline src/faces/FacePage.vala]: Indendation error

File: src/faces/FacePage.vala [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        return groups;
+    }
+
+    private InjectionGroup create_faces_menu_injectables(){
+        InjectionGroup menuFaces = new InjectionGroup("FacesMenuPlaceholder");
+       
+        //menuFaces.add_menu("Faces");
+        menuFaces.add_menu_item("Remove Face \"" + this.face.get_name() + "\" From Photos", "RemoveFaceFromPhotos", "<Primary>r");
+        menuFaces.add_menu_item("Rename Face \"" + this.face.get_name() + "\"…", "RenameFace", "<Primary>e");
+        menuFaces.add_menu_item("Dele_teFace", "DeleteFace", "<Primary>t");
```
Comment (phako) [inline src/faces/FacePage.vala]: Missing white space, also texts should be translatable

File: src/library/LibraryWindow.vala [hunk 13]
```diff
@@ -269,6 +279,148 @@ public class LibraryWindow : AppWindow {
         monitor.auto_import_progress.disconnect(on_library_monitor_auto_import_progress);
     }
 
+/*<<<<<<< HEAD
+        Gtk.ActionEntry preferences = { "CommonPreferences", Resources.PREFERENCES_LABEL, TRANSLATABLE,
```
Comment (phako) [inline src/library/LibraryWindow.vala]: Do you really need to keep this all in here?

File: src/meson.build [hunk 14]
```diff
@@ -17,13 +17,24 @@ processor = executable('shotwell-graphics-processor',
                        dependencies: [gio, gdk, gee],
                        link_with: sw_graphics_processor)
 
+face_resources = []
```
Comment (phako) [inline src/meson.build]: Technically, that's face_sources; resources is used for the resource files (ui, icons etc.)

**Review by phako — CHANGES_REQUESTED** [review]
Some things need to be cleaned up. After that, I think we can merge it (and after the feature freeze). But the faces stuff definitely needs some more work after that

**Comment by ricardofantin** [comment]
Thank you for the comments. Saturday I should make all changes.

In my next changes I will take more care about tabs and comments.

**Comment by ricardofantin** [comment]
Hello.
I made almost everything you asked, the only thing left is to correctly write the code to replace the deprecated stock buttons. I also noted that are many places with code to use the faces. To make the code better I should reduce the places, besides the src/faces/ folder, which have faces related code. In some files there are 4 different places with faces related code. I will try to reduce it. After removing the code from src/editing_tools/EditingTools.vala and creating src/faces/FacesTool.vala there are more 13 files with faces related code. It will take a while to get the code better.
The "-Dface=true" were taken from the Makefile from the old faces code. I chose to use "-Dface-detection=true".
I have many questions, if you could answer some of them would really help me:
1. Which IDE you use to edit vala? (I'm used vim and devhelp, and now I'm using geany)
2. To make a text translatable, I can make an entry in the ui/faces.ui or put the string in src/Resource.vala, right? Is there another way? I'm almost sure that the better solution is make the entry in ui/faces.ui, is it right?
3. Instead of using the Gtk.Button.from_stock(Gtk.Stock.EDIT); I'm creating a new Button and taking the text from Resources.EDIT_LABEL, but the button appear with label "_Edit" . There should not be the underline before the Edit. How do I remove it and make the system use the translated text? Another thing we should discuss is the icons, some icons from the stock buttons are not available through icon name. So which icons I should use in the buttons? The new icons and the text with underlines are attached. The old icons used are attached too, the text in this version is in Brazilian Portuguese, but only look at the icons.
old-icons-from-stockbutton
_edit-and-icones
4. In the copyright header, can I put the year of 2018?
5. I made the facedetect/shotwell-facedetect.cpp copyright header this way:

```
/* Copyright 2016 Software Freedom Conservancy Inc.
 * 
 * Copyright 2011 Valentín Barros Puertas <valentin(at)sanva(dot)net>
 * Copyright 2018 Ricardo Fantin da Costa <ricardofantin(at)gmail(dot)com>
 * 
 * This software is licensed under the GNU LGPL (version 2.1 or later).
 * See the COPYING file in this distribution.
 */
```
Is this copyright ok?

Thank you for your time. I'll try to make the better code I can.

**Comment by phako** [comment]
1. I use vim or gnome-builder occasionally
2. Yes, please put it in the ui file where possible. If you use strings in vala code, plesae make sure to prefix it with _, e.g. _("Some text to translate")
3. In Glade, check the "Always use underline" checkbox. In code, use new Button.with_mnemonic() or set use_underline to true
4. Yes, use 2018. The header also looks fine

**Comment by phako** [comment]
1. I use vim or gnome-builder occasionally
2. Yes, please put it in the ui file where possible. If you use strings in vala code, plesae make sure to prefix it with _, e.g. _("Some text to translate")
3. In Glade, check the "Always use underline" checkbox. In code, use new Button.with_mnemonic() or set use_underline to true. For the icon use "gtk-edit" - that's basically cheating, but I don't know any better solution for now
4. Yes, use 2018. The header also looks fine

**Comment by ricardofantin** [comment]
Hello.
All requested changes were made.
The commit message resumes all major changes. Some minor changes include substitution uncorrected tabs used in indentation to spaces, remove of many commented code.

I start make the Makefile.ac in an old commit (Not included is this push). I want to finish this code to the use can compile in two ways. And make a name autocompletion to finish the bug [717733](https://bugzilla.gnome.org/show_bug.cgi?id=717733) .

Some things to make in the future:
In src/faces/FacesTool.vala:
faces = new Gee.PriorityQueue<string>();
"As a backlog item, we should move this to something that's slightly nicer to parse than this"

In many files there are Faces related code in different places, maybe put everything just in one place.

Any feedback, is welcome.

Bye.

**Comment by phako** [comment]
Hi, sorry, i was busy. Please forget about Makefile.ac. It is gone on master. I will do a deeper check tomorrow

**Comment by phako** [comment]
You forgot to add src/faces/FacesTool.vala to the commit

**Comment by ricardofantin** [comment]
I'll make a force amend to include src/faces/FacesTool.vala and will update the meson.build version too.
Just a minute.

**Comment by ricardofantin** [comment]
Everything correct now.

File: src/faces/FacesTool.vala [hunk 15]
```diff
@@ -0,0 +1,971 @@
...
+                    if (cancel_detection_button.get_parent() == null)
+                        help_layout.pack_start(cancel_detection_button, false);
+
+                    detection_button.set_sensitive(false);
+                    cancel_detection_button.set_sensitive(true);
+                    cancel_detection_button.show();
+
+                    break;
+                case EditingPhase.DETECTING_FACES_FINISHED:
+                    help_text.set_text(_("If you don't set the name of unknown faces they won't be saved."));
```
Comment (phako) [inline src/faces/FacesTool.vala]: use unicode apostrophe ’ inste’ad of normal '

File: src/faces/FacesTool.vala [hunk 16]
```diff
@@ -0,0 +1,971 @@
...
+                string[] argv = {
+                    AppDirs.get_facedetect_bin().get_path(),
+                    "--cascade=" + AppDirs.get_haarcascade_file().get_path(),
+                    "--scale=1.2",
+                    image_path
+                };
+                Process.spawn_sync(null, argv, null, SpawnFlags.STDERR_TO_DEV_NULL, null, out output);
+
+            } catch (SpawnError e) {
+                stderr.printf("Error trying to spawn face detection program: %s\n", e.message);
```
Comment (phako) [inline src/faces/FacesTool.vala]: Make that a criticla() log message. Since this is a background job, save the error, then in the finished callback, check for the error and show it to the user with AppWindow.error()

File: src/faces/FacesTool.vala [hunk 17]
```diff
@@ -0,0 +1,971 @@
...
+
+            faces = new Gee.PriorityQueue<string>();
+            string[] lines = output.split("\n");
+            foreach (string line in lines) {
+                if (line.length == 0)
+                    continue;
+
+                string[] type_and_serialized = line.split(";");
+                if (type_and_serialized.length != 2) {
+                    stderr.printf("Wrong serialized line in face detection program output.");
```
Comment (phako) [inline src/faces/FacesTool.vala]: use critical() instead of stderr.printf. Assert can stay, though

File: src/faces/FacesTool.vala [hunk 18]
```diff
@@ -0,0 +1,971 @@
...
+                        StringBuilder serialized_geometry = new StringBuilder();
+                        serialized_geometry.append(FaceRectangle.SHAPE_TYPE);
+                        serialized_geometry.append(";");
+                        serialized_geometry.append(parse_serialized_geometry(type_and_serialized[1]));
+
+                        faces.add(serialized_geometry.str);
+                        break;
+
+                    case "warning":
+                        stderr.printf("%s\n", type_and_serialized[1]);
```
Comment (phako) [inline src/faces/FacesTool.vala]: warning()

File: src/faces/FacesTool.vala [hunk 19]
```diff
@@ -0,0 +1,971 @@
...
+                        break;
+
+                    case "error":
+                        stderr.printf("%s\n", type_and_serialized[1]);
```
Comment (phako) [inline src/faces/FacesTool.vala]: critical(), as above, save error, show to user

File: src/faces/FacesTool.vala [hunk 20]
```diff
@@ -0,0 +1,971 @@
...
+
+            double x = 0;
+            double y = 0;
+            double width = 0;
+            double height = 0;
+            foreach (string piece in serialized_geometry_pieces) {
+
+                string[] name_and_value = piece.split("=");
+                if (name_and_value.length != 2) {
+                    stderr.printf("Wrong serialized line in face detection program output.");
```
Comment (phako) [inline src/faces/FacesTool.vala]: critical()

**Review by phako — CHANGES_REQUESTED** [review]
Please fix those, then we are good to go.

**Comment by phako** [comment]
The very first task after the merge should be fixing the horrid user experience of editing the faces names. It is completely annoying that a small window with just an entry pops up

**Comment by ricardofantin** [comment]
Everything corrected.

I looked how Digicam show faces, and soon I will make a better interface to define the person's name.

File: src/faces/FacesTool.vala [hunk 21]
```diff
@@ -337,7 +337,7 @@ public class FacesTool : EditingTools.EditingTool {
                 Process.spawn_sync(null, argv, null, SpawnFlags.STDERR_TO_DEV_NULL, null, out output);
 
             } catch (SpawnError e) {
-                stderr.printf("Error trying to spawn face detection program: %s\n", e.message);
+                critical("Error trying to spawn face detection program: %s\n", e.message);
                 assert_not_reached();
```
Comment (phako) [inline src/faces/FacesTool.vala]: You missed one comment here. This is not a programming error, hence I would not assert.
I would rather remember that there was an error in the FaceDetectionJob, exit the function with "return", then check in the callback if the Job had an error and present it to the user with AppWindow.error_message

**Comment by ricardofantin** [comment]
Good night.

I finally finish the error message when, for any reason, the program can't call shotwell-facedetect.
To see the error message compile and after delete facedetect/shotwell-facedetect .

The annoying floating entry is really bad.  I think something like digikam should be ok:
digikam-editname

The entry is with the main window, and not in a float. When the user use the detect faces it should be good change the label in facedetect window to entries as the user clicks.

Bye.

File: src/faces/FacesTool.vala [hunk 22]
```diff
@@ -337,8 +337,10 @@ public class FacesTool : EditingTools.EditingTool {
                 Process.spawn_sync(null, argv, null, SpawnFlags.STDERR_TO_DEV_NULL, null, out output);
 
             } catch (SpawnError e) {
-                critical("Error trying to spawn face detection program: %s\n", e.message);
-                assert_not_reached();
+                output = "Error trying to spawn face detection program: " + e.message + "\n";
+                stdout.printf("Erro: %s", output);
+                critical(output);
```
Comment (phako) [inline src/faces/FacesTool.vala]: You reintroduced a printf here

File: src/faces/FacesTool.vala [hunk 23]
```diff
@@ -950,8 +952,12 @@ public class FacesTool : EditingTools.EditingTool {
...
     private void on_faces_detected() {
         face_detection_cancellable.reset();
-
-        pick_faces_from_autodetected();
+        
+        if ("Error trying to spawn face detection program: " in face_detection.output){
+            AppWindow.error_message(face_detection.output);
+            faces_tool_window.set_editing_phase(EditingPhase.DETECTING_FACES_FINISHED);
+		}else
+            pick_faces_from_autodetected();
```
Comment (phako) [inline src/faces/FacesTool.vala]: Please fix the code style, that indentation looks off

File: src/faces/FacesTool.vala [hunk 24]
```diff
@@ -337,8 +337,10 @@ public class FacesTool : EditingTools.EditingTool {
                 Process.spawn_sync(null, argv, null, SpawnFlags.STDERR_TO_DEV_NULL, null, out output);
 
             } catch (SpawnError e) {
-                critical("Error trying to spawn face detection program: %s\n", e.message);
-                assert_not_reached();
+                output = "Error trying to spawn face detection program: " + e.message + "\n";
```
Comment (phako) [inline src/faces/FacesTool.vala]: If this is shown to the user, it needs to be marked translatable. Why don't you save the SpawnError directly? Then you only have to do a != null check below.

**Review by phako — CHANGES_REQUESTED** [review]

**Comment by ricardofantin** [comment]
Hello,
the last three errors were correct.

I can't wait to, after the merge, start asking to make the face-detection option to be default to true.

I have a deadline in next day 28, after that I'll have time to make the user interface changes.

Bye.

**Comment by phako** [comment]
There are some minor style issues which I will fix while merging. I will merge this now.

For the GUI enhancements, you can use a PR on Shotwell's new GNOME Gitlab instance https://gitlab.gnome.org/GNOME/shotwell

**Comment by phako** [comment]
Pushed to master
